### ISSUE TITLE
Error when trying to assign role to user

### ISSUE THREAD

**Issue by achanda** [mattupstate/flask-security#661]
I am trying to use flask-security to implement RBAC. This is how I am trying to use the signals to assign roles to users (using SQLAlchemy)

```
@user_registered.connect_via(app)
def user_registered_sighandler(sender, **extra):
    user  = extra.get('user')
    admin_role = user_datastore.find_or_create_role(name='admin', description='Administrator')
    user_role = user_datastore.find_or_create_role(name='user', description='User')
    user_datastore.add_role_to_user(user, admin_role)
```

This gives me this error
```
Traceback (most recent call last):
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/flask/app.py", line 1997, in __call__
    return self.wsgi_app(environ, start_response)
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/flask/app.py", line 1985, in wsgi_app
    response = self.handle_exception(e)
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/flask_cors/extension.py", line 161, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/flask/app.py", line 1540, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/flask/_compat.py", line 33, in reraise
    raise value
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/flask/app.py", line 1982, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/flask/app.py", line 1614, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/flask_cors/extension.py", line 161, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/flask/app.py", line 1517, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/flask/_compat.py", line 33, in reraise
    raise value
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/flask/app.py", line 1612, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/flask/app.py", line 1598, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/flask_security/decorators.py", line 230, in wrapper
    return f(*args, **kwargs)
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/flask_security/views.py", line 118, in register
    user = register_user(**form.to_dict())
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/flask_security/registerable.py", line 37, in register_user
    user=user, confirm_token=token)
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/blinker/base.py", line 267, in send
    for receiver in self.receivers_for(sender)]
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/blinker/base.py", line 267, in <listcomp>
    for receiver in self.receivers_for(sender)]
  File "/Users/Abhishek/src/datasine/canopy/canopy/routes/__init__.py", line 78, in user_registered_sighandler
    user_datastore.add_role_to_user(user, admin_role)
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/flask_security/datastore.py", line 156, in add_role_to_user
    user.roles.append(role)
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/sqlalchemy/orm/collections.py", line 1045, in append
    item = __set(self, item, _sa_initiator)
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/sqlalchemy/orm/collections.py", line 1017, in __set
    item = executor.fire_append_event(item, _sa_initiator)
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/sqlalchemy/orm/collections.py", line 680, in fire_append_event
    item, initiator)
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/sqlalchemy/orm/attributes.py", line 951, in fire_append_event
    initiator or self._append_token or self._init_append_token())
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/sqlalchemy/orm/unitofwork.py", line 47, in append
    sess._save_or_update_state(item_state)
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 1691, in _save_or_update_state
    self._save_or_update_impl(state)
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 1997, in _save_or_update_impl
    self._update_impl(state)
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 1980, in _update_impl
    to_attach = self._before_attach(state, obj)
  File "/Users/Abhishek/anaconda3/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 2064, in _before_attach
    state.session_id, self.hash_key))
sqlalchemy.exc.InvalidRequestError: Object '<Role at 0x109f45a58>' is already attached to session '2' (this is '3')
```
I tried running a `db_session.commit()` right after doing the role assignment which had the same problem. What am I doing wrong?

**Comment by jirikuncar** [comment]
The SQLAlchemy session is already committed.

1. try to create new session in your signal handler and get user and roles again
2. there might be a possible issue with missing explicit `begin` method on datastore

```python
# please check that you are using correct datastore in your handler
datastore = sender.extensions['flask-security'].datastore
```

**Comment by achanda** [comment]
Thanks @jirikuncar, but I got a similar error with this
```
@user_registered.connect_via(app)
def user_registered_sighandler(sender, **extra):
    engine = create_engine(os.environ.get('DATABASE_URL'), pool_size=64, max_overflow=64)
    session = scoped_session(sessionmaker(autocommit=False, autoflush=False, bind=engine))
    user  = extra.get('user')
    admin_role = user_datastore.find_or_create_role(name='admin', description='Administrator')
    user_role = user_datastore.find_or_create_role(name='user', description='User')
    user_datastore.add_role_to_user(user, admin_role)
    session.commit()
```
The stacktrace this time looks similar with the error being
```
sqlalchemy.exc.InvalidRequestError: Object '<Role at 0x1135e3860>' is already attached to session '1' (this is '2')
```
I am using SQLAlchemy
```
(Pdb) sender.extensions['security'].datastore
<flask_security.datastore.SQLAlchemyUserDatastore object at 0x11037c240>
```

**Comment by jirikuncar** [comment]
@achanda have you tried to define `user_datastore` as `user_datastore = sender.extensions['flask-security'].datastore`? I see the problem in accessing user and roles after commit without refreshing their state. I don't know how to solve the problem in generic way for all supported datastores, but in this case I would try also http://docs.sqlalchemy.org/en/latest/orm/session_state_management.html#expunging or http://docs.sqlalchemy.org/en/latest/orm/session_state_management.html#refreshing-expiring `user`, `admin_role` and `user_role`.

**Comment by achanda** [comment]
@jirikuncar just tried that and it did not seem to work. Here is my code for reference https://gist.github.com/achanda/6b20434651fcfd1ed95ecbaf0b8f16bd

Here is everything I tried:
1. Creating a new session in `user_registered_sighandler` using `sessionmaker`
2. Getting a new reference to `user_datastore` in `user_registered_sighandler` using `sender.extensions['security'].datastore`. The key seems to be `security` and not `flask-security`. Is that a problem?
3. Manually calling `db_session.expunge_all()` in the sig handler before anything else.
4. Manually expiring `admin_role` and `user_role`

All of those had the same error. Do you happen to see anything obviously wrong here?

**Comment by achanda** [comment]
@jirikuncar I found the problem (partially). I was creating the `db_session` object using `sqlalchemy` directly. In the code I posted earlier, I was creating another db object using `Flask-SQLAlchemy`. If I try to pass `db_session` directly to `SQLAlchemyUserDatastore`, I get this error
```
AttributeError: 'scoped_session' object has no attribute 'session'
```
Any suggestions how can I make this work without using `Flask-SQLAlchemy`?
